modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/AbstractSortedIndexStorageTest.java

```diff
@@ -1233,6 +1235,276 @@ void testScanContractRemoveNextAndAddFirstRow() {
         assertThrows(NoSuchElementException.class, scan::next);
     }
 
+
+    @Test
+    void testScanPeekForFinishedCursor() {
+        SortedIndexDefinition indexDefinition = SchemaBuilders.sortedIndex("TEST_IDX")
+                .addIndexColumn(ColumnType.INT32.typeSpec().name()).asc().done()
+                .build();
+
+        SortedIndexStorage indexStorage = createIndexStorage(indexDefinition);
+
+        BinaryTupleRowSerializer serializer = new BinaryTupleRowSerializer(indexStorage.indexDescriptor());
+
+        PeekCursor<IndexRow> scan0 = indexStorage.scan(null, null, 0);
+        PeekCursor<IndexRow> scan1 = indexStorage.scan(null, null, 0);
+
+        // index   =
+        // cursor0 = ^ already finished
+        assertFalse(scan0.hasNext());
+        assertNull(scan0.peek());
+
+        // index   =
+        // cursor1 = ^ already finished
+        assertThrows(NoSuchElementException.class, scan1::next);
+        assertNull(scan1.peek());
+
+        // index   =  [0]
+        // cursor0 = ^ already finished
+        // cursor1 = ^ already finished
+        put(indexStorage, serializer.serializeRow(new Object[]{0}, new RowId(TEST_PARTITION)));
+
+        assertNull(scan0.peek());
+        assertNull(scan1.peek());
+
+        // index   =  [0]
+        // cursor0 = ^ no cached row
+        // cursor1 = ^ no cached row
+        scan0 = indexStorage.scan(null, null, 0);
+        scan1 = indexStorage.scan(null, null, 0);
+
+        assertEquals(0, serializer.deserializeColumns(scan0.peek())[0]);
+        assertEquals(0, serializer.deserializeColumns(scan1.peek())[0]);
+
+        // index   = [0]
+        // cursor0 =    ^ cached [0]
+        assertTrue(scan0.hasNext());
```

Comment:
This is very strange: why does calling `hasNext` invalidate what `peek` returns? I thought only `next` can do that

Reply:
I will fix